core/include/scipp/core/view_index.h

```diff
@@ -5,61 +5,62 @@
 #pragma once
 
 #include "scipp-core_export.h"
+#include "scipp/common/index_composition.h"
 #include "scipp/core/dimensions.h"
+#include "scipp/core/strides.h"
 
 namespace scipp::core {
 
 class SCIPP_CORE_EXPORT ViewIndex {
 public:
-  ViewIndex(const Dimensions &targetDimensions,
-            const Dimensions &dataDimensions);
+  ViewIndex(const Dimensions &target_dimensions, const Strides &strides);
 
   constexpr void increment_outer() noexcept {
     scipp::index d = 0;
-    while ((m_coord[d] == m_extent[d]) && (d < NDIM_MAX - 1)) {
-      m_index += m_delta[d + 1];
+    while ((m_coord[d] == m_shape[d]) && (d < NDIM_MAX - 1)) {
+      m_memory_index += m_delta[d + 1];
       ++m_coord[d + 1];
       m_coord[d] = 0;
       ++d;
     }
   }
   constexpr void increment() noexcept {
-    m_index += m_delta[0];
+    m_memory_index += m_delta[0];
     ++m_coord[0];
-    if (m_coord[0] == m_extent[0])
+    if (m_coord[0] == m_shape[0])
       increment_outer();
-    ++m_fullIndex;
+    ++m_view_index;
   }
 
-  constexpr void setIndex(const scipp::index index) noexcept {
-    m_fullIndex = index;
-    if (m_dims == 0)
-      return;
-    auto remainder{index};
-    for (int32_t d = 0; d < m_dims - 1; ++d) {
-      if (m_extent[d] != 0) {
-        m_coord[d] = remainder % m_extent[d];
-        remainder /= m_extent[d];
-      } else {
-        m_coord[d] = 0;
-      }
+  constexpr void set_index(const scipp::index index) noexcept {
+    m_view_index = index;
+    extract_indices(index, m_ndim, m_shape, m_coord);
+    m_memory_index = flat_index_from_strides(
+        m_strides.begin(), m_strides.end(m_ndim), m_coord.begin());
```

Comment:
Have you compared performance to the old approach? It was always very slow and should not be relied on, but wondering if it got any better or worse?

Reply:
No. Is there already a benchmark for this?

Reply:
Hmm, no, according to a quick look through our benchmarks. I think you simply need to check
```python
auto vals = var.values<double>();
vals[i];
```
for some `i`.

Reply:
There is `element_array_view_benchmark.cpp` which would cover the non-random access, I think?

Reply:
So you are explicitly asking about the performance of `set_index`?

Reply:
Hmm, I suppose both would be good to know, `operator[]()` as well as iteration/incrementing?